pkg/storageops/gce/gce.go

```diff
@@ -0,0 +1,540 @@
+
+const googleDiskPrefix = "/dev/disk/by-id/google-"
+
+type gceOps struct {
+	inst    *instance
+	service *compute.Service
+	mutex   sync.Mutex
+}
+
+// instance stores the metadata of the running GCE instance
+type instance struct {
+	ID         string
+	Name       string
+	Hostname   string
+	Zone       string
+	Project    string
+	InternalIP string
+	ExternalIP string
+	LBRequest  string
+	ClientIP   string
+	Error      string
+}
+
+type assigner struct {
+	err error
+}
+
+func (a *assigner) assign(getVal func() (string, error)) string {
+	if a.err != nil {
+		return ""
+	}
+
+	s, err := getVal()
+	if err != nil {
+		a.err = err
+	}
+
+	return s
+}
+
+func getEnvValueStrict(key string) (val string, err error) {
+	if val = os.Getenv(key); len(val) != 0 {
+		return
+	}
+
+	err = fmt.Errorf("env variable %s is not set", key)
+	return
+}
+
+// gceInfo fetches the GCE instance metadata from the metadata server
+func gceInfo(inst *instance) {
+	a := &assigner{}
+	inst.ID = a.assign(metadata.InstanceID)
+	inst.Zone = a.assign(metadata.Zone)
+	inst.Name = a.assign(metadata.InstanceName)
+	inst.Hostname = a.assign(metadata.Hostname)
+	inst.Project = a.assign(metadata.ProjectID)
+	inst.InternalIP = a.assign(metadata.InternalIP)
+	inst.ExternalIP = a.assign(metadata.ExternalIP)
+
+	if a.err != nil {
+		inst.Error = a.err.Error()
+	}
+}
+
+func gceInfoFromEnv(inst *instance) {
+	a := &assigner{}
+	inst.Name = a.assign(func() (string, error) { return getEnvValueStrict("GCE_INSTANCE_NAME") })
+	inst.Zone = a.assign(func() (string, error) { return getEnvValueStrict("GCE_INSTANCE_ZONE") })
+	inst.Project = a.assign(func() (string, error) { return getEnvValueStrict("GCE_INSTANCE_PROJECT") })
+
+	if a.err != nil {
+		inst.Error = a.err.Error()
+	}
+}
+
+// IsDevMode checks if the pkg is invoked in developer mode where GCE credentials
+// are set as env variables
+func IsDevMode() bool {
+	var i = new(instance)
+	gceInfoFromEnv(i)
+	return len(i.Error) == 0
+}
+
+// NewClient creates a new GCE operations client
+func NewClient() (storageops.Ops, error) {
+	var i = new(instance)
+	if metadata.OnGCE() {
+		gceInfo(i)
+	} else if ok := IsDevMode(); ok {
+		gceInfoFromEnv(i)
+	} else {
+		return nil, fmt.Errorf("instance is not running on GCE")
+	}
+
+	if len(i.Error) != 0 {
+		return nil, fmt.Errorf("error while fetching instance information. Err: %v", i.Error)
+	}
+
+	c, err := google.DefaultClient(context.Background(), compute.ComputeScope)
+	if err != nil {
+		return nil, fmt.Errorf("failed to authenticate with google api. Err: %v", err)
+	}
+
+	service, err := compute.New(c)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create Compute service: %v", err)
+	}
+
+	return &gceOps{
+		inst:    i,
+		service: service,
+	}, nil
+}
+
+func (s *gceOps) waitStatus(id string, desired string) error {
+	actual := ""
+	for retries, maxRetries := 0, storageops.ProviderOpsMaxRetries; actual != desired && retries < maxRetries; retries++ {
```

Comment:
Take a look at https://github.com/kubernetes/apimachinery/blob/master/pkg/util/wait/wait.go . We should instead have a `util` pkg in osd, so that it can be used by other packages.

Reply:
This does not have to be done for this PR. If you want, please create an issue that solves this and add it here.

Reply:
We already had https://github.com/portworx/sched-ops/blob/master/task/task.go#L15

So I have now changed all retry loops to use this.